src/herder/ParallelTxSetBuilder.cpp

```diff
@@ -434,87 +580,60 @@ buildSurgePricedParallelSorobanPhase(
+                                 cfg.SOROBAN_PHASE_MIN_STAGE_COUNT + 1;
+    std::vector<ParallelPhaseBuildResult> results(stageCountOptions);
 
-    std::vector<Resource> laneLeftUntilLimitUnused;
-    queue.popTopTxs(/* allowGaps */ true, visitor, laneLeftUntilLimitUnused,
-                    hadTxNotFittingLane);
-    // There is only a single fee lane for Soroban, so there is only a single
-    // flag that indicates whether there was a transaction that didn't fit into
-    // lane (and thus all transactions are surge priced at once).
-    releaseAssert(hadTxNotFittingLane.size() == 1);
-
-    // At this point the stages have been filled with transactions and we just
-    // need to place the full transactions into the respective stages/clusters.
-    TxStageFrameList resStages;
-    resStages.reserve(stages.size());
-    for (auto const& stage : stages)
+    for (uint32_t stageCount = cfg.SOROBAN_PHASE_MIN_STAGE_COUNT;
+         stageCount <= cfg.SOROBAN_PHASE_MAX_STAGE_COUNT; ++stageCount)
     {
-        auto& resStage = resStages.emplace_back();
-        resStage.reserve(partitionCfg.mClustersPerStage);
-
-        std::unordered_map<size_t, size_t> clusterIdToStageCluster;
-
-        stage.visitAllTransactions([&resStage, &txFrames,
-                                    &clusterIdToStageCluster](size_t clusterId,
-                                                              size_t txId) {
-            auto it = clusterIdToStageCluster.find(clusterId);
-            if (it == clusterIdToStageCluster.end())
-            {
-                it = clusterIdToStageCluster.emplace(clusterId, resStage.size())
-                         .first;
-                resStage.emplace_back();
-            }
-            resStage[it->second].push_back(txFrames[txId]);
+        size_t resultIndex = stageCount - cfg.SOROBAN_PHASE_MIN_STAGE_COUNT;
+        threads.emplace_back([queue, &builderTxForTx, txFrames, stageCount,
+                              sorobanCfg, laneConfig, resultIndex, &results]() {
+            results[resultIndex] =
+                buildSurgePricedParallelSorobanPhaseWithStageCount(
+                    std::move(queue), builderTxForTx, txFrames, stageCount,
+                    sorobanCfg, laneConfig);
         });
-        // Algorithm ensures that clusters are populated from first to last and
-        // no empty clusters are generated.
-        for (auto const& cluster : resStage)
-        {
-            releaseAssert(!cluster.empty());
-        }
     }
-    // Ensure we don't return any empty stages, which is prohibited by the
-    // protocol. The algorithm builds the stages such that the stages are
-    // populated from first to last.
-    while (!resStages.empty() && resStages.back().empty())
+    for (auto& thread : threads)
     {
-        resStages.pop_back();
+        thread.join();
     }
-    for (auto const& stage : resStages)
+
+    releaseAssert(!results.empty());
```

Comment:
The size should be `SOROBAN_PHASE_MAX_STAGE_COUNT` right? Should we assert on that?

Reply:
Yeah, I'm not sure what I have been thinking about here. Removed.